File: examples/02-reliability/go/parallel_fanout_allof.go
package reliability

import (
	"time"

	"go.temporal.io/sdk/workflow"
)

// Launch one Activity per partition, collect the Futures, then Get each to sum
// the counts. Collecting before getting is what makes them run in parallel.
func ProcessPartitions(ctx workflow.Context, partitions []int) (int, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 15 * time.Minute,
	})

	futures := make([]workflow.Future, 0, len(partitions))
	for _, partition := range partitions {
		futures = append(futures, workflow.ExecuteActivity(ctx, "ProcessPartition", partition))
	}

	total := 0
	for _, f := range futures {
		var count int
		if err := f.Get(ctx, &count); err != nil {
			return 0, err
		}
		total += count
	}
	return total, nil
}

File: examples/02-reliability/python/parallel_fanout_allof.py
import asyncio
from datetime import timedelta

from temporalio import workflow

# `process_partition` is a module-level @activity.defn function returning an int.


@workflow.defn
class PartitionFanoutWorkflow:
    # Launch one Activity per partition, then asyncio.gather to wait for all of
    # them. gather preserves order, so summing the results is straightforward.
    @workflow.run
    async def process_partitions(self, partitions: list[int]) -> int:
        counts = [
            workflow.execute_activity(
                process_partition,
                partition,
                start_to_close_timeout=timedelta(minutes=15),
            )
            for partition in partitions
        ]
        return sum(await asyncio.gather(*counts))
